### ISSUE TITLE
Symbols

### ISSUE THREAD

**Issue by Ic3MaN8187** [sspanak/tt9#647]
The lack of  importan symbols  and how to find then is terrible. Backspace button is in the wrong place, should be where (CLR) is. (. And ,) should have a dedicated key. Extremely confusing and zero intuitive layout.. touchpal is the best example to copy!

**Comment by sspanak** [comment]
You mean copy a keyboard that supports T9 only in some languages, that misuses the file permission to write random files in different places in your phone, and that got banned from the Play Store? I don't think so. 

If you like Touchpal so much, just use it. I will never turn Traditional T9 into it.

And there is no need of hate posts.

**Comment by Ic3MaN8187** [comment]
First, it is the best T9 keyboard ever, nothing can't touch him  second, I was saying copying the layout and how practical he was, and with the best dictionaries I've ever seen..not the spying part.. yours is extremely confusing  and pretty much unusable.. not hating just the truth!

**Comment by sspanak** [comment]
Ok.

**Comment by sspanak** [comment]
Since I am getting a lot of hate from some former Touchpal users (or employees?), I've decided to keep this open. I have also added a special section in the manual where you can find [a comparison between the two apps](https://github.com/sspanak/tt9/blob/master/docs/help/help.en.md#you-should-copy-touchpal-its-the-greatest-keyboard-in-the-world) and why TT9 is not and will not become Touchpal. 

Bottom line is, adapt or use some other app.

**Comment by maverick74** [comment]
What's the point of keeping it open, may i ask?

**Comment by sspanak** [comment]
I am getting a lot of comments on Google Play, not here. I think it is better to keep it open, both for new users and for search engines to find it easier. I hope this will reduce the spam a bit...

**Comment by mauricewanjala82-pixel** [comment]
Actually TouchPal is the best example to copy. Look, we have swipe input or gesture input, contextual sentence prediction, auto correction of spelling, auto prediction, quick  delete, etc.

**Comment by mauricewanjala82-pixel** [comment]
One major reason why I use tt9 is because it supports Swahili language, which is not supported in TouchPal t9 layout. Otherwise for English typing, TouchPal t9 layout is easier to type. Could TouchPal support Swahili in t9 layout, then it could have been very far from traditional t9 keyboard. TouchPal is very powerful and has qualities to be copied. Traditional t9 is simpler with some good qualities.

**Comment by mauricewanjala82-pixel** [comment]
Kiswahili dictionary should be expanded

**Comment by maverick74** [comment]
> we have swipe input or gesture input,

I, personally, never understood the _hype_ with this!!! Dragging your finger across the screen seems "counter productive"... but that's me!!! I already understood there's a market for this... i just don't like it!

(edit: oh and... btw, I think the correct form is _had_ - in the past - not _have_! TouchPal is dead afaik... there are a few fake copies but you should get away from those)

> contextual sentence prediction, auto prediction

I believe we already have something like this but sspanak can correct me here (I think we currently consider 2 words but may be expand it to 3 at the cost of CPU)

> quick delete

What exactly do you mean with this? If you swipe left on the delete key it will erase the last word as Touchpal did! That's quick delete... Do you mean something else? If so, can you elaborate?

> auto correction of spelling

Yeah... I wanted that too...  we've talked about this a lot before at #525 and it turns out it's quite difficult to accomplish!
TouchPal was probably using cloud (as in: sending in what you type to get analysed and used to provide correction predictions) and I believe no one wants to go down that path (I'm sure I don't) so... while "amazing" i don't see how it can be done anytime soon...


> Otherwise for English typing, TouchPal t9 layout is easier to type.

Easier how? What are your points of view on this? Where is it better? Any other points than the ones previously mentioned?
I used TouchPal for more than 10 years and, aside from auto correction (that had that _cloud detail_), there is actually nothing else I can think of where touchpal is better! Quite the contrary Iyam... can you elaborate?

**Comment by sspanak** [comment]
Wow, it looks like I missed the party! I remember @mauricewanjala82-pixel. He complained on the Play Store about missing features that are actually there. He's just a hater who has not explored all the possibilities of TT9. Anyway, here are my answers.

> Actually TouchPal is the best example to copy.

Really? But how do I use that TouchPal thing? It doesn't type anything, just numbers.

https://github.com/user-attachments/assets/0fbb3114-ba73-4df4-86fa-5ba5cd2c7f8f

> One major reason why I use tt9 is because it supports Swahili language

Let's stop at TouchPal does not work. It does not work on newer Android versions, it does not support hardware keyboards or keypads, it does not support TV remote controls, it does not support that many languages. And, in my humble opinion, a keyboard that does not work, is far from being the best. I would say it is the worst.

> Look, we have swipe input or gesture input, contextual sentence prediction, auto correction of spelling, auto prediction, quick delete, etc.

**Swipe and gesture input** is useful on QWERTY keyboards, where one key produces only one letter. When typing using 9 keys, very often it is necessary to press the same key multiple times, which is mutually exclusive with swiping. For example, how do you swipe "mono" (6-6-6-6) or "back" (2-2-2-4), or "high" (4-4-4-4), and so on? 

Moreover, on touchscreen devices, TT9 is meant to recreate the classic Nokia experience where you press the keys. And on devices with hardware keys, naturally, you can't really swipe. 

For these reasons, I will not make any efforts to add swipe typing. Period.

**Contextual sentence prediction** is supported and it actually works better than TouchPal. Personally, I find it extremely annoying when I attempt to type "I am an...", I get "am" twice in a row, instead of getting "an" when preceded by "am". Another example is "the tie". When I try to type this, all keyboards, TouchPal included, would suggest "the", "the" twice. TT9 does not do this. The only deficiency is, it needs to learn the word combinations from you, which takes some time. But, if you type actively it should be mostly OK in a day. Also, I intend to improve the predictions and even include an option to predict words before you have started typing. See #851 for more info. @maverick74, I think I already have a good idea how to do it at minimal CPU cost.

**Auto-correction** is a nice thing but, unfortunately, I lack the knowledge to do it. As @maverick74 pointed out, see the discussion #525.

**Quick-delete** is available but is disabled by default. Read the manual and explore the app before complaining that it is missing.

> Kiswahili dictionary should be expanded

How do you think this works? You request stuff and it magically appears in the app? I don't speak the language and I don't have any friends from Tanzania, so I can't do more than what I have already done. I have no idea where to search for more language-related resources. 

I you want the app to improve, contribute, don't just come here and give me orders. Provide a richer word list, and I would gladly add it to the current dictionary.

And one last thing...

> TouchPal is the best example to copy. Look, we have... 

You made me wonder, what do you mean by "we have"? Who are "we"? Are you a former TouchPal employee or you are like those fanatic football fans who think they belong to the team and say "we" when they refer to it? In both cases, why should I care about your "great" keyboard? Just stay with it. I never asked you to use TT9.

**Comment by maverick74** [comment]
> Also, I intend to improve the predictions and even include an option to **_predict words before you have started typing_**. See [#851](https://github.com/sspanak/tt9/issues/851) for more info. [@maverick74](https://github.com/maverick74), I think **I already have a good idea how to do it at _minimal CPU cost_**.

Sweet music to my ears 🎉

**Comment by mauricewanjala82-pixel** [comment]
OK. 
TT9 is a good keyboard. That is why I give out suggestions on what could be added on it. 

Swipe/gesture input 
Predict words before you have started typing 
Auto correction.
Swahili dictionary expansion.
Wide space bar. 

On t9 keyboard layout, when you press a key, a letter is printed. When you swipe on a key downward, a number is printed. This is what I referred to it as gesture or swipe input. Example on a key with abc has a number 2.
When you press directly on a key with letters a b c, you type a letter a with suggestion b c.. When you swipe downward on that key, you type a number 2.  This was my suggested idea. I think this saves time while typing  letters and numbers without switching to numberpad.

If possible space bar could be made wider. This could reduce errors where by instead of pressing space bar, you press another key (! And?) 

Swahili words,especially verbs in( present, past, future) tense, In singular and plural. If possible I may provide a list of them. 


Predicting words before you have started typing them. Example when you type the letter I then you press space bar, suggested words or predicted words like am, have, etc comes up. 

Auto correction. Suppose you did not master the spelling of the word. This is tricky. Correcting word from wrongly pressed keys. 

 I have been giving reference on TouchPal keyboard especially on features which I was used to.

**Comment by mauricewanjala82-pixel** [comment]
When you type wrong spelling like "inicially" should auto corrected as "initially"

**Comment by maverick74** [comment]
> On t9 keyboard layout, when you press a key, a letter is printed. When you swipe on a key downward, a number is printed. This is what I referred to it as gesture or swipe input. Example on a key with abc has a number 2. When you press directly on a key with letters a b c, you type a letter a with suggestion b c.. When you swipe downward on that key, you type a number 2. This was my suggested idea. I think this saves time while typing letters and numbers without switching to numberpad.

You do know that if you hold down the abc key you get the number 2, don't you??? Because "type and hold" and "swipe down" will take about the same time... it's just a question of getting used to it!!!

> If possible space bar could be made wider. This could reduce errors where by instead of pressing space bar, you press another key (! And?)

You do know there is an option to make space bar wider already, don't you??? I mean... you did take a look to the app preferences and you actually want an even wider (like an edge to edge) space bar, right?!?!

Because if not... you're just wasting dev time...

> (... ) all the other requests

You did noticed sspanak's (aka tt9's DEV) reply above about all the other issues you mention...???

In short (again!):

Predict words: **IS COMING**!!!

Auto correction: **IS _NOT_ COMING ANYTIME SOON - if ever...**
(unless you know any _magic_ trick that unlocks the knowledge on how to do that with minimal CPU cost and you share it with the dev)

Swahili dictionary expansion: **JUST CONTRIBUTE AND IT WILL GET BETTER**

**Comment by sspanak** [comment]
> When you type wrong spelling like "inicially" should auto corrected as "initially"

That's a general problem with T9 typing. You must know how to spell, in order to type the words. If you press the wrong key, you don't just get "inicially", you get an entirely different word. And from there, is extremely difficult to deduce what you intended to type. I don't have a solution for this.

**Comment by maverick74** [comment]
> I don't have a solution for this.

That's the only thing where touchpal still has an edge.... the detail is "_the price_" we would have to pay for that _edge_ ...

(I'm not willing to quit my privacy for this feature, nor having tt9 considerably more slow...)

**Comment by NeonLogicHub** [comment]
I am surprised at the level of patience, the staff has had in this thread regarding the other keyboard. Lol I get both sides. Long time touchpal user. Bought a pixel fold with Android 16 and had to switch to another keyboard. TT9 is almost at a point that if you configure it properly it's almost the same. As for the one's that complain, they should be thankful we have an alternative and a very responsive dev team. And also.... Donate or buy a subscription to show support! I did.

**Comment by sspanak** [comment]
Thanks for the support! Actually, keeping this issue open has reduced the spam significantly. I guess people do read it.

And just to clarify, there is no "team" or "staff", it's just me doing most of the work in my free time (full time in the past year), and occasionally, enthusiast would join briefly, to fix a bug, or add their native language.

**Comment by NeonLogicHub** [comment]
> Thanks for the support! Actually, keeping this issue open has reduced the spam significantly. I guess people do read it.
> 
> And just to clarify, there is no "team" or "staff", it's just me doing most of the work in my free time (full time in the past year), and occasionally, enthusiast would join briefly, to fix a bug, or add their native language.

It comes from the fact that for years, touchpal has been the only keyboard that allowed people to use  true T9, just like in the old days. As long as one would block the Internet access the data couldn be collected nor transmitted. 

And then it no longer works after android 13 and people start searching for an alternative and here we are. The sad fact is that  like on a lot of other open source projects, people are entitled and on top of it they Don't donate. Remember reading a lengthy article about it once it a tech website....

**Comment by NeonLogicHub** [comment]
> Thanks for the support! Actually, keeping this issue open has reduced the spam significantly. I guess people do read it.
> 
> And just to clarify, there is no "team" or "staff", it's just me doing most of the work in my free time (full time in the past year), and occasionally, enthusiast would join briefly, to fix a bug, or add their native language.

Hey man! 

I have been thinking about this for quite sometime, but i assume it's best to ask rather than just posting. Are you ok with me making a "small guide here" on how to configure TT9 to best resemble Touchpal?

I say this because as a paid subscriber, i was able to combine the best of both worlds. TT9 and the aforementioned keyboard. I still had it installed on my oneplus 8 pro (2020 phone), and now on my pixel 10 pro fold with TT9 , i replicated what was possible on the design layout and then took advantage of the amazing configuration options that TT9 offers. But i am not sure it's cool to share here. Let me know your thoughts.

**Comment by sspanak** [comment]
Please, go ahead. Also, feel free to publish your guide on social media, forums and whatnot. From the feedback I receive, it looks like many people don't even realize how many configuration options TT9 provides.

**Comment by maverick74** [comment]
I believe that posted here would also be very valuable!!!

I'm interested in seeing it as well!!!
